proof/verifier.go

```diff
@@ -184,67 +188,273 @@ func (p *Proof) verifySplitRootProof() error {
 func (p *Proof) verifyExclusionProofs() (*commitment.TapCommitmentVersion,
 	error) {
 
-	// Gather all P2TR outputs in the on-chain transaction.
-	p2trOutputs := make(map[uint32]struct{})
+	// There is nothing to check so return early.
+	if p.ExclusionProofs == nil {
+		return nil, nil
+	}
+
+	// Gather all P2TR outputs in the on-chain transaction. The STXO proofs
+	// are tracked using a set of serialized keys. We ignore that set when
+	// verifying V0 exclusion proofs.
+	p2trOutputs := make(fn.Set[uint32])
 	for i, txOut := range p.AnchorTx.TxOut {
 		if uint32(i) == p.InclusionProof.OutputIndex {
 			continue
 		}
-		if txscript.IsPayToTaproot(txOut.PkScript) {
-			p2trOutputs[uint32(i)] = struct{}{}
+		if !txscript.IsPayToTaproot(txOut.PkScript) {
+			continue
 		}
+
+		p2trOutputs.Add(uint32(i))
 	}
 
-	// Verify all of the encoded exclusion proofs.
-	commitVersions := make(map[uint32]commitment.TapCommitmentVersion)
-	for _, exclusionProof := range p.ExclusionProofs {
-		exclusionProof := exclusionProof
-		derivedCommitment, err := verifyTaprootProof(
-			&p.AnchorTx, &exclusionProof, &p.Asset, false,
+	// Early pass to determine if we're dealing with v0 or v1 proofs.
+	hasV1Proofs := len(p.ExclusionProofs) > 0 &&
+		p.ExclusionProofs[0].CommitmentProof != nil &&
+		len(p.ExclusionProofs[0].CommitmentProof.STXOProofs) > 0
+
+	if hasV1Proofs {
+		return p.verifyV1ExclusionProofs(p2trOutputs)
+	}
+
+	return p.verifyV0ExclusionProofs(p2trOutputs)
+}
+
+// verifyV0ExclusionProofs verifies all version 1 exclusion proofs.
+func (p *Proof) verifyV0ExclusionProofs(p2trOutputs fn.Set[uint32]) (
+	*commitment.TapCommitmentVersion, error) {
+
+	commitVersions := make(map[uint32][]commitment.TapCommitmentVersion)
+
+	// Verify all of the encoded v0 exclusion proofs.
+	for idx := range p.ExclusionProofs {
+		exclusionProof := p.ExclusionProofs[idx]
+		err := p.handleBasicExclusionProof(
+			&exclusionProof, p2trOutputs, commitVersions,
 		)
 		if err != nil {
 			return nil, err
 		}
-
-		outputIdx := exclusionProof.OutputIndex
-		delete(p2trOutputs, outputIdx)
-
-		// Store the commitment version. If there was no Taproot Asset
-		// commitment present, then there is nothing to store.
-		if derivedCommitment != nil {
-			commitVersions[outputIdx] = derivedCommitment.Version
-		}
 	}
 
 	// If any outputs are missing a proof, fail.
 	if len(p2trOutputs) > 0 {
 		return nil, ErrMissingExclusionProofs
 	}
 
-	// If there were no commitments in any exclusion proofs, then there is
-	// no version to return.
+	// If no commitments in proofs, no version to return.
+	if len(commitVersions) == 0 {
+		return nil, nil
+	}
+
+	// All proofs must have similar versions.
+	firstVersion := maps.Values(commitVersions)[0][0]
```

Comment:
Move this extraction of the first version into the `verifySTXOVersions`, since it's done the same way twice?